substrate/primitives/runtime/src/traits/transaction_extension.rs

```diff
@@ -0,0 +1,671 @@
+		len: usize,
+		result: &DispatchResult,
+	) -> Result<(), TransactionValidityError> {
+		for_tuples!( #( Tuple::post_dispatch(pre.Tuple, info, post_info, len, result)?; )* );
+		Ok(())
+	}
+
+	fn metadata() -> Vec<TransactionExtensionMetadata> {
+		let mut ids = Vec::new();
+		for_tuples!( #( ids.extend(Tuple::metadata()); )* );
+		ids
+	}
+}
+
+impl<Call: Dispatchable> TransactionExtension<Call> for () {
+	const IDENTIFIER: &'static str = "UnitTransactionExtension";
+	type Val = ();
+	type Pre = ();
+	type Implicit = ();
+	fn implicit(&self) -> sp_std::result::Result<Self::Implicit, TransactionValidityError> {
+		Ok(())
+	}
+	fn validate(
+		&self,
+		origin: <Call as Dispatchable>::RuntimeOrigin,
+		_call: &Call,
+		_info: &DispatchInfoOf<Call>,
+		_len: usize,
+		_implicit: &[u8],
+	) -> Result<
+		(ValidTransaction, (), <Call as Dispatchable>::RuntimeOrigin),
+		TransactionValidityError,
+	> {
+		Ok((ValidTransaction::default(), (), origin))
+	}
+	fn prepare(
+		self,
+		_val: (),
+		_origin: &<Call as Dispatchable>::RuntimeOrigin,
+		_call: &Call,
+		_info: &DispatchInfoOf<Call>,
+		_len: usize,
+	) -> Result<(), TransactionValidityError> {
+		Ok(())
+	}
+}
+
+/// Adapter to use a `SignedExtension` in the place of a `TransactionExtension`.
+#[derive(TypeInfo, Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
+//#[deprecated = "Convert your SignedExtension to a TransactionExtension."]
+pub struct AsTransactionExtension<SE: SignedExtension>(pub SE);
+
+impl<SE: SignedExtension + Default> Default for AsTransactionExtension<SE> {
+	fn default() -> Self {
+		Self(SE::default())
+	}
+}
+
+impl<SE: SignedExtension> From<SE> for AsTransactionExtension<SE> {
+	fn from(value: SE) -> Self {
+		Self(value)
+	}
+}
+
+impl<SE: SignedExtension> TransactionExtension<SE::Call> for AsTransactionExtension<SE>
+where
+	<SE::Call as Dispatchable>::RuntimeOrigin: CloneSystemOriginSigner<SE::AccountId> + Clone,
+{
+	const IDENTIFIER: &'static str = SE::IDENTIFIER;
+	type Val = ();
+	type Pre = SE::Pre;
+	type Implicit = SE::AdditionalSigned;
+
+	fn implicit(&self) -> Result<Self::Implicit, TransactionValidityError> {
+		self.0.additional_signed()
+	}
+
+	fn validate(
+		&self,
+		origin: <SE::Call as Dispatchable>::RuntimeOrigin,
+		call: &SE::Call,
+		info: &DispatchInfoOf<SE::Call>,
+		len: usize,
+		_implicit: &[u8],
+	) -> Result<
+		(ValidTransaction, (), <SE::Call as Dispatchable>::RuntimeOrigin),
+		TransactionValidityError,
+	> {
+		let who = origin.clone_system_origin_signer().ok_or(InvalidTransaction::BadSigner)?;
+		Ok((self.0.validate(&who, call, info, len)?, (), origin))
+	}
+
+	fn prepare(
+		self,
+		_: (),
+		origin: &<SE::Call as Dispatchable>::RuntimeOrigin,
+		call: &SE::Call,
+		info: &DispatchInfoOf<SE::Call>,
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		let who = origin.clone_system_origin_signer().ok_or(InvalidTransaction::BadSigner)?;
+		self.0.pre_dispatch(&who, call, info, len)
```

Comment:
Hm we dont really need to clone in either function.  
Something like `as_system_origin_signer(&self) -> Option<&AccountId>` should suffice. I can change it once you are done with the bulk of changes.